client/allocrunner/taskrunner/identity_hook.go

```diff
@@ -151,3 +189,121 @@ func (h *identityHook) getIdentities(alloc *structs.Allocation, task *structs.Ta
 
 	return widMap, nil
 }
+
+// renew fetches new signed workload identity tokens before the existing tokens
+// expire.
+func (h *identityHook) renew(createIndex uint64, signedWIDs map[string]*structs.SignedWorkloadIdentity) {
+	wids := h.task.Identities
+	if len(wids) == 0 {
+		h.logger.Trace("no workload identities to renew")
+		return
+	}
+
+	var reqs []*structs.WorkloadIdentityRequest
+	renewNow := false
+	minExp := time.Now().Add(30 * time.Hour)                        // set high default expiration
+	widMap := make(map[string]*structs.WorkloadIdentity, len(wids)) // Identity.Name -> Identity
+
+	for _, wid := range wids {
+		if wid.TTL == 0 {
+			// No ttl, so no need to renew it
+			continue
+		}
+
+		widMap[wid.Name] = wid
+
+		reqs = append(reqs, &structs.WorkloadIdentityRequest{
+			AllocID:      h.alloc.ID,
+			TaskName:     h.task.Name,
+			IdentityName: wid.Name,
+		})
+
+		sid, ok := signedWIDs[wid.Name]
+		if !ok {
+			// Missing a signature, treat this case as already expired so we get a
+			// token ASAP
+			h.logger.Trace("missing token for identity", "identity", wid.Name)
+			renewNow = true
+			continue
+		}
+
+		if sid.Expiration.Before(minExp) {
+			minExp = sid.Expiration
+		}
+	}
+
+	if len(reqs) == 0 {
+		h.logger.Trace("no workload identities expire")
+		return
+	}
+
+	var wait time.Duration
+	if !renewNow {
+		wait = helper.ExpiryToRenewTime(minExp, time.Now, h.minWait)
+	}
+
+	timer, timerStop := helper.NewStoppedTimer()
+	defer timerStop()
+
+	var retry uint64
+
+	for err := h.stopCtx.Err(); err == nil; {
+		h.logger.Trace("waiting to renew identities", "num", len(reqs), "wait", wait)
+		timer.Reset(wait)
+		select {
+		case <-timer.C:
+			h.logger.Debug("getting new signed identities", "num", len(reqs))
+		case <-h.stopCtx.Done():
+			return
+		}
+
+		// Renew all tokens together since its cheap
+		tokens, err := h.widmgr.SignIdentities(createIndex, reqs)
+		if err != nil {
+			retry++
+			wait = helper.Backoff(h.minWait, time.Hour, retry) + helper.RandomStagger(h.minWait)
+			h.logger.Error("error renewing workload identities", "error", err, "next", wait)
+			continue
+		}
+
+		if len(tokens) == 0 {
+			retry++
+			wait = helper.Backoff(h.minWait, time.Hour, retry) + helper.RandomStagger(h.minWait)
+			h.logger.Error("error renewing workload identities", "error", "no tokens", "next", wait)
+			continue
+		}
+
+		// Reset next expiration time
+		minExp = time.Time{}
+
+		for _, token := range tokens {
+			widspec, ok := widMap[token.IdentityName]
+			if !ok {
+				// Bug: Every requested workload identity should either have a signed
+				// identity.
```

Comment:
"should either have a signed identity" or...? I think they all get signed identities in 1.7.0 going forward, even existing allocs across the upgrade?

Reply:
`...or rejection.` was the intended completion of that sentence! Good catch, fixed.